### PR TITLE
Avoid MakeGenericMethod in XmlSerializer

### PR DESCRIPTION
**Pull request by MichalStrehovsky**
I pulled out the `IsValueType` part of the generic method into the non-generic method since it doesn't use anything from the generic context. This should be an improvement irrespective of AOT.

Then I made us take the IsValueType path if `IsDynamicCodeSupported` is false. This has two consequences:

* Avoids `MakeGenericType` that may or may not succeed with `IsDynamicCodeSupported=false`
* Avoids LINQ expressions in the field case. This is will be a throughput improvement.
* There might be a small throughput regression for the property case when IsDynamicCodeSupported is false, but we get reliability in return.

Cc @dotnet/ilc-contrib

### CODE REVIEW

File: src/libraries/System.Private.Xml/src/System/Xml/Serialization/ReflectionXmlSerializationReader.cs [hunk 1]
```diff
@@ -635,24 +635,44 @@ private static ReflectionXmlSerializationReaderHelper.SetMemberValueDelegate Get
...
-                            memberType = propInfo.PropertyType;
-                        }
-                        else if (memberInfo is FieldInfo fieldInfo)
+
+                        if (type.IsValueType || !RuntimeFeature.IsDynamicCodeSupported)
                         {
-                            memberType = fieldInfo.FieldType;
+                            if (memberInfo is PropertyInfo propInfo)
+                            {
+                                result = new ReflectionXmlSerializationReaderHelper.SetMemberValueDelegate(propInfo.SetValue);
```
Comment (eerhardt) [inline src/libraries/System.Private.Xml/src/System/Xml/Serialization/ReflectionXmlSerializationReader.cs:643]: For my learning, is there a difference between:

```C#
 new ReflectionXmlSerializationReaderHelper.SetMemberValueDelegate(propInfo.SetValue);
```

and

```C#
propInfo.SetValue.CreateDelegate<ReflectionXmlSerializationReaderHelper.SetMemberValueDelegate>();
```

?

**Comment by eerhardt** [comment]
@dotnet/area-system-xml @HongGit @StephenMolloy @mconnew - any feedback here?

File: src/libraries/System.Private.Xml/src/System/Xml/Serialization/ReflectionXmlSerializationReader.cs [hunk 2]
```diff
@@ -2097,46 +2117,30 @@ internal static class ReflectionXmlSerializationReaderHelper
 
         public static SetMemberValueDelegate GetSetMemberValueDelegateWithType<TObj, TParam>(MemberInfo memberInfo)
         {
-            if (typeof(TObj).IsValueType)
```
Comment (eerhardt) [inline src/libraries/System.Private.Xml/src/System/Xml/Serialization/ReflectionXmlSerializationReader.cs:2100]: Would it make sense to add `Debug.Assert(!typeof(TObj).IsValueType);`? Just to document that we don't expect any structs to come into this method.

**Review by eerhardt — APPROVED** [review]

**Review by StephenMolloy — APPROVED** [review]
lgtm

**Pull request merged**